src/domain/tradingview/useTVDatafeed.ts

```diff
@@ -86,108 +95,195 @@ export default function useTVDatafeed({ dataProvider }: Props) {
-              });
-            }
-          };
-
-          if (!isStable) {
-            intervalRef.current = setInterval(handleInterval, 500);
-          }
-        },
-        unsubscribeBars: (id) => {
-          // id is in the format ETH_#_USD_#_5
-          const ticker = id.split("_")[0];
-          const isStable = stableTokens.includes(ticker);
-          if (!isStable && intervalRef.current) {
-            clearInterval(intervalRef.current);
-          }
-        },
+          const noData = !bars || bars.length === 0;
+          onHistoryCallback(bars, { noData });
+        } catch {
+          onErrorCallback("Unable to load historical data!");
+        }
       },
-    };
-  }, [chainId, stableTokens, supportedResolutions]);
+      async subscribeBars(
+        symbolInfo: SymbolInfo,
+        resolution: ResolutionString,
+        onRealtimeCallback: SubscribeBarsCallback
+      ) {
+        await subscribeBars({
+          symbolInfo,
+          resolution,
+          onRealtimeCallback,
+          chainId,
+          missingBarsInfoRef,
+          feedDataRef,
+          lastBarTimeRef,
+          intervalRef,
+          supportedResolutions,
+          tvDataProviderRef,
+        });
+      },
+      unsubscribeBars: (id) => {
+        // id is in the format ETH_#_USD_#_5
+        const ticker = id.split("_")[0];
+        const isStable = stableTokens.includes(ticker);
+        if (!isStable && intervalRef.current) {
+          clearInterval(intervalRef.current);
+        }
+      },
+    },
+  };
+}
+
+export function subscribeBars({
+  symbolInfo,
+  resolution,
+  onRealtimeCallback,
+  chainId,
+  feedDataRef,
+  intervalRef,
+  supportedResolutions,
+  tvDataProviderRef,
+  lastBarTimeRef,
+  missingBarsInfoRef,
+}: {
+  symbolInfo: Pick<SymbolInfo, "ticker" | "isStable">;
+  resolution: ResolutionString;
+  onRealtimeCallback: SubscribeBarsCallback;
+  chainId: number;
+  supportedResolutions: { [key: string]: string };
+  tvDataProviderRef: MutableRefObject<TVDataProvider | undefined>;
+  intervalRef: MutableRefObject<number | undefined>;
+  missingBarsInfoRef: MutableRefObject<{
+    bars: FromOldToNewArray<Bar>;
+    isFetching: boolean;
+  }>;
+  feedDataRef: MutableRefObject<boolean>;
+  lastBarTimeRef: MutableRefObject<number>;
+}) {
+  const period = supportedResolutions[resolution];
+  const { ticker, isStable } = symbolInfo;
+  if (!ticker) {
+    return;
+  }
+
+  if (intervalRef.current !== undefined) {
+    clearInterval(intervalRef.current);
+    intervalRef.current = undefined;
+  }
+
+  const handleInterval = () => {
+    if (missingBarsInfoRef.current.isFetching || !feedDataRef.current) return;
+    if (missingBarsInfoRef.current.bars?.length > 0) {
+      let processedBarTimes: FromOldToNewArray<number> = [];
+      missingBarsInfoRef.current.bars.forEach((bar: Bar, index) => {
+        if (processedBarTimes.includes(bar.time)) return;
+        if (index !== 0 && bar.time < processedBarTimes.at(-1)!) {
```

Comment:
could you explain?
so `missingBarsInfoRef.current.bars` should go in descending order? can't we just sort them?

Reply:
1) We sort them multiple times throughout the pipeline. Would not want to add another sorting. The problem was in the order the `onRealtimeCallback` was called
2) ascending.